lib/config/convict.js

```diff
@@ -0,0 +1,835 @@
+    const url = new URL(val)
+    if (!/^https?:/.test(url.protocol)) {
+      throw new Error('Not a valid HTTP URL!')
+    }
+  }
+})
+
+convict.addFormat({
+  name: 'semver',
+  validate: function(val) {
+    if (!/^\d+\.\d+\.\d+([-+].*)?$/.test(val)) {
+      throw new Error('Not valid SemVer! See https://semver.org for details.')
+    }
+  }
+})
+
+// Define a schema
+let config = convict({
+  debug: {
+    doc: 'Enable debug mode for CodiMD.',
+    format: 'Boolean',
+    env: 'DEBUG',
+    default: false
+  },
+  env: {
+    doc: 'The application environment.',
+    format: ['production', 'development', 'test'],
+    default: 'development',
+    env: 'NODE_ENV'
+  },
+  domain: {
+    doc: 'Domain used for CodiMD. Without this set, some functionality might be broken.',
+    format: function (val) {
+      if (val !== undefined && !domainRegex.test(val)) {
+        throw new Error("Domain not valid, please provide a proper domain name. For example: example.com or codimd.example.com")
+      }
+    },
+    env: 'CMD_DOMAIN',
+    default: undefined
+  },
+  urlPath: {
+    doc: 'Sub-URL-path, like `example.com/<urlpath>`',
+    format: 'String',
+    env: 'CMD_URL_PATH',
+    default: ''
+  },
+  host: {
+    doc: 'IP address CodiMD should bind to. 0.0.0.0 and :: will be interpreted as all IPs of the host.',
+    format: 'ipaddress',
+    env: 'CMD_HOST',
+    default: '0.0.0.0'
+  },
+  port: {
+    doc: 'Port CodiMD should bind to.',
+    format: 'port',
+    env: 'CMD_PORT',
+    default: 3000
+  },
+  path: {
+    doc: 'UNIX socket path to listen on, as alternative to regular TCP socket listening',
+    format: 'optional-path',
+    env: 'CMD_PATH',
+    default: undefined
+  },
+  loglevel: {
+    doc: 'Defines what kind of logs are provided to stdout. Available options: `debug`, `verbose`, `info`, `warn`, `error`',
+    format: function (val) {
+      if (!/(debug|verbose|info|warn|error)/.test(val)) {
+        throw new Error('Loglevel has to be set to `debug`, `verbose`, `info`, `warn`, or `error`')
+      }
+    },
+    env: 'CMD_LOGLEVEL',
+    default: 'info'
+  },
+  urlAddPort: {
+    doc: 'set to add port on callback URL (ports `80` or `443` won\'t be applied) (only applied when domain is set)',
+    format: 'Boolean',
+    env: 'CMD_URL_ADDPORT',
+    default: false
+  },
+  allowOrigin: {
+    doc: 'CORS whitelist for non-API calls',
+    format: function (val) {
+      for(let i = val.length; i--;) {
+        if (!domainRegex.test(val[i])) {
+          throw new Error(`CORS array contains invalid domain name '${val[i]}'`)
+        }
+      }
+    },
+    env: 'CMD_ALLOW_ORIGIN',
+    default: ['localhost']
+  },
+  useSSL: {
+    doc: 'Set to use SSL server (if `true`, will auto turn on `protocolUseSSL`',
+    format: 'Boolean',
+    env: 'CMD_USESSL',
+    default: false
+  },
+  hsts: {
+    enable: {
+      doc: 'Set to enable [HSTS](https://en.wikipedia.org/wiki/HTTP_Strict_Transport_Security) if HTTPS is also enabled',
```

Comment:
does markdown make sense here? at a quick glance, I didn't find anything on markdown in the node-convict documentation.

Reply:
Yeah, it's rather questionable. It was also mainly copy-paste from the docs.

Reply:
do we / can we generate markdown files from the configuration? if so, i'd leave it in. If not, i'd go for 
```suggestion
      doc: 'Set to enable HTTP Strict Transport Security if HTTPS is also enabled. For more info visit https://en.wikipedia.org/wiki/HTTP_Strict_Transport_Security',
```